wallet/src/main/java/ch/admin/bag/covidcertificate/wallet/transfercode/worker/TransferWorker.kt

```diff
@@ -0,0 +1,119 @@
+ *
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+package ch.admin.bag.covidcertificate.wallet.transfercode.worker
+
+import android.content.Context
+import androidx.work.*
+import ch.admin.bag.covidcertificate.common.config.ConfigModel
+import ch.admin.bag.covidcertificate.wallet.BuildConfig
+import ch.admin.bag.covidcertificate.wallet.data.WalletDataItem
+import ch.admin.bag.covidcertificate.wallet.data.WalletDataSecureStorage
+import ch.admin.bag.covidcertificate.wallet.transfercode.logic.TransferCodeCrypto
+import ch.admin.bag.covidcertificate.wallet.transfercode.model.TransferCodeModel
+import ch.admin.bag.covidcertificate.wallet.transfercode.net.DeliveryRepository
+import ch.admin.bag.covidcertificate.wallet.transfercode.net.DeliverySpec
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import java.io.IOException
+import java.util.concurrent.TimeUnit
+
+class TransferWorker(context: Context, workerParams: WorkerParameters) : CoroutineWorker(context, workerParams) {
+
+	companion object {
+		val WORKER_NAME = TransferWorker::class.java.canonicalName
+
+		private const val DEFAULT_REPEAT_INTERVAL = 120 * 60 * 1000L
+		private const val DEFAULT_FLEX_INTERVAL = 5 * 60 * 1000L
+		private const val DEFAULT_BACKOFF_DELAY = 30 * 1000L
+
+		fun scheduleTransferWorker(context: Context, config: ConfigModel? = null) {
+			val transferWorkRequest = PeriodicWorkRequest.Builder(
+				TransferWorker::class.java,
+				DEFAULT_REPEAT_INTERVAL,
+				TimeUnit.MILLISECONDS,
+				DEFAULT_FLEX_INTERVAL,
+				TimeUnit.MILLISECONDS
+			)
+				.setBackoffCriteria(BackoffPolicy.EXPONENTIAL, DEFAULT_BACKOFF_DELAY, TimeUnit.MILLISECONDS)
+				.build()
+
+			WorkManager.getInstance(context)
+				.enqueueUniquePeriodicWork(WORKER_NAME, ExistingPeriodicWorkPolicy.KEEP, transferWorkRequest)
+		}
+
+		fun cancelScheduledTransferWorker(context: Context) {
+			WorkManager.getInstance(context).cancelUniqueWork(WORKER_NAME)
+		}
+	}
+
+	private val walletDataStorage = WalletDataSecureStorage.getInstance(context)
+	private val deliveryRepository = DeliveryRepository.getInstance(DeliverySpec(context, BuildConfig.BASE_URL_DELIVERY))
+
+	private val workManager = WorkManager.getInstance(context)
+
+	override suspend fun doWork(): Result {
+		return withContext(Dispatchers.IO) {
+			val transferStates = walletDataStorage.getWalletData()
+				.filterIsInstance<WalletDataItem.TransferCodeWalletData>()
+				.map { transferCodeItem -> downloadTransferCertificate(transferCodeItem.transferCode) }
+				.toSet()
+
+			if (transferStates.contains(TransferState.ERROR)) {
+				return@withContext Result.retry()
+			} else if (!transferStates.contains(TransferState.NOT_AVAILABLE)) {
+				workManager.cancelUniqueWork(WORKER_NAME)
+			}
+			Result.success()
+		}
+	}
+
+	private suspend fun downloadTransferCertificate(transferCode: TransferCodeModel): TransferState {
+		val keyPair = TransferCodeCrypto.loadKeyPair(transferCode.code)
+
+		if (keyPair != null) {
```

Comment:
Personally I quite like the style of doing a check and early return:
```kotlin
if (keyPair == null) {
 return ERROR
}
```
This eliminates the need for an else block, resulting in less indentation and better readability.